packages/@aws-cdk/aws-s3/lib/bucket.ts

```diff
@@ -1502,6 +1512,11 @@ export class Bucket extends BucketBase {
       throw new Error(`encryptionKey is specified, so 'encryption' must be set to KMS (value: ${encryptionType})`);
     }
 
+    // if bucketKeyEnabled is set, encryption must be set to KMS.
+    if (props.bucketKeyEnabled !== undefined && encryptionType !== BucketEncryption.KMS) {
```

Comment:
My initial reasoning would be to say yes, because 

- If I set `bucketKeyEnabled` to `false` explicitly and it just works I would wonder why - leaving everything else the same - if set to `true` it wouldn't.

For me it is not clean to say "You can set this to false disregarding X, but you can only set it to true if X".

Reply:
> If I set bucketKeyEnabled to false explicitly and it just works I would wonder why - leaving everything else the same - if set to true it wouldn't.

Not sure what you mean, why would i wonder? I explicitly set it to `false`, meaning i want to ignore any limitations/validations derived from it. Setting it back to `true` is the user indicating the opposite. 

Also, creating a distinction between `undefined`, `true`, and `false` with booleanic values is confusing, especially for languages like `Java`, where boolean types are assigned `false` by default. 

If `false` and `undefined` are functionally equivalent, we should be explicit with `false`. If not, maybe a boolean is not the right type. (enum?)

Reply:
Your argument makes sense. Will implement it this way